src/ClientGenerator/AppDomainCodeGenerator.cs

```diff
@@ -0,0 +1,121 @@
+using System.IO;
+using System.Reflection;
+using Microsoft.Extensions.Logging;
+using Orleans.CodeGeneration;
+using Orleans.Runtime.Configuration;
+using Orleans.CodeGenerator;
+using Orleans.Serialization;
+using Orleans.Logging;
+
+#if NET461
+namespace Orleans.CodeGeneration
+{
+    public class AppDomainCodeGenerator : MarshalByRefObject
+    {
+        public static string GenerateCode(CodeGenOptions options)
+        {
+            AppDomain appDomain = null;
+            try
+            {
+                var assembly = typeof(AppDomainCodeGenerator).GetTypeInfo().Assembly;
+
+                // Create AppDomain.
+                var appDomainSetup = new AppDomainSetup
+                {
+                    ApplicationBase = Path.GetDirectoryName(assembly.Location),
+                    DisallowBindingRedirects = false,
+                    ConfigurationFile = AppDomain.CurrentDomain.SetupInformation.ConfigurationFile
+                };
+                appDomain = AppDomain.CreateDomain("Orleans-CodeGen Domain", null, appDomainSetup);
+
+                // Set up assembly resolver
+                var refResolver = new ReferenceResolver(options.ReferencedAssemblies);
+                appDomain.AssemblyResolve += refResolver.ResolveAssembly;
+
+                // Create an instance 
+                var generator =
+                    (AppDomainCodeGenerator)
+                    appDomain.CreateInstanceAndUnwrap(
+                        assembly.FullName,
+                        // ReSharper disable once AssignNullToNotNullAttribute
+                        typeof(AppDomainCodeGenerator).FullName);
+
+                // Call a method 
+                return generator.GenerateCodeInternal(options);
+            }
+            finally
+            {
+                if (appDomain != null) AppDomain.Unload(appDomain); // Unload the AppDomain
+            }
+        }
+
+        private string GenerateCodeInternal(CodeGenOptions options)
+        {
+            // Load input assembly 
+            // special case Orleans.dll because there is a circular dependency.
+            var assemblyName = AssemblyName.GetAssemblyName(options.InputAssembly.FullName);
+            var grainAssembly = Path.GetFileName(options.InputAssembly.FullName) != "Orleans.dll"
+                ? Assembly.LoadFrom(options.InputAssembly.FullName)
+                : Assembly.Load(assemblyName);
+
+            // Create directory for output file if it does not exist
+            var outputFileDirectory = Path.GetDirectoryName(options.OutputFileName);
+
+            if (!String.IsNullOrEmpty(outputFileDirectory) && !Directory.Exists(outputFileDirectory))
+            {
+                Directory.CreateDirectory(outputFileDirectory);
+            }
+
+            var config = new ClusterConfiguration();
+            var loggerFactory = new LoggerFactory();
+            loggerFactory.AddProvider(new FileLoggerProvider("ClientGenerator.log"));
```

Comment:
BTW, this is putting a ClientGenerator.log alonside the code. It should be in the intermediate folder, correct?

Reply:
Yeah, I ported this over from ClientGenerator in master, but I didn't like it (because of that file being generated) so I'm using ConsoleLoggerProvider now, which has required pulling in a new package.